### PR TITLE
Move oc_share to migration

### PR DESCRIPTION
**Pull request by PVince81**
## Description
Move the oc_share table to be created through a migration instead of db_structure.xml for the future.

## Related Issue
See https://github.com/owncloud/core/pull/27337

## Motivation and Context
To make it possible to add new columns (and maybe even delete if we decide to cleanup) more easily as we require a migration for that, so better have the whole table in a migration instead.

## How Has This Been Tested?
Diffed the MySQL schemas between master and this branch, both were identical.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@DeepDiver1975 @butonic @jvillafanez please review.

The real reason is that we want to add a "name" column for multiple link shares to be able to visually distinguish them in the UI.

### CODE REVIEW

File: core/Migrations/Version20170315173825.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+class Version20170315173825 implements ISchemaMigration {
+
+	public function changeSchema(Schema $schema, array $options) {
+		$prefix = $options['tablePrefix'];
+		if (!$schema->hasTable("${prefix}share")) {
+			$table = $schema->createTable("${prefix}share");
+			$table->addColumn('id', 'integer', [
+				'autoincrement' => true,
+				'notnull' => true,
+				'length' => 4,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: schema uses a default value, but I guess it isn't needed

File: core/Migrations/Version20170315173825.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+use OCP\Migration\ISchemaMigration;
+
+/**
+ * Initial DB creation for share table
+ */
+class Version20170315173825 implements ISchemaMigration {
+
+	public function changeSchema(Schema $schema, array $options) {
+		$prefix = $options['tablePrefix'];
+		if (!$schema->hasTable("${prefix}share")) {
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: Taking into account the indentation level is still acceptable and that no improvements are expected in this file, there is no need to change this for the indentation.

File: core/Migrations/Version20170315173825.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+			$table->addColumn('id', 'integer', [
+				'autoincrement' => true,
+				'notnull' => true,
+				'length' => 4,
+			]);
+			// Constant OCP\Share::SHARE_TYPE_*
+			$table->addColumn('share_type', 'smallint', [
+				'notnull' => true,
+				'default' => 0,
+				'length' => 6,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: length was 1 before, not sure if we want to change it.

File: core/Migrations/Version20170315173825.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			]);
+			$table->addColumn('file_target', 'string', [
+				'notnull' => false,
+				'length' => 512,
+			]);
+			// Permission bitfield
+			$table->addColumn('permissions', 'smallint', [
+				'default' => 0,
+				'notnull' => true,
+				'length' => 6,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: different length here

File: core/Migrations/Version20170315173825.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			]);
+			// Time of share creation
+			$table->addColumn('stime', 'bigint', [
+				'default' => 0,
+				'notnull' => true,
+				'length' => 20,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: different length

File: core/Migrations/Version20170315173825.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+			]);
+			// Whether the receiver accepted the share, if share_with is set.
+			$table->addColumn('accepted', 'smallint', [
+				'default' => 0,
+				'notnull' => true,
+				'length' => 6,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: different length

File: core/Migrations/Version20170315173825.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+			]);
+			// Public share token
+			$table->addColumn('token', 'string', [
+				'notnull' => false,
+				'length' => 32,
+			]);
+			$table->addColumn('mail_send', 'smallint', [
+				'default' => 0,
+				'notnull' => true,
+				'length' => 6,
```
Comment (jvillafanez) [inline core/Migrations/Version20170315173825.php]: different length

**Comment by jvillafanez** [comment]
The data types are also different, but I guess it will be handled properly and the value range for the integers fit fine.

I'm not sure if the length value for the integers is used, maybe we can remove it to avoid confusion.

**Comment by PVince81** [comment]
I'm not sure about the sizes, they seem correct as the DB diff was exactly the same.

From what I heard the old mdb2 code used some nasty tricks with the sizes so they might not be the same now in Doctrine than they were before. @DeepDiver1975 might be able to confirm.

**Comment by PVince81** [comment]
@jvillafanez I talked with @DeepDiver1975 and apparently we can get rid of the lengths altogether for numeric values. Here is how the past did it: https://github.com/owncloud/core/blob/v9.1.4/lib/private/DB/MDB2SchemaReader.php#L260

I'll remove the lengths and check the diffs again.

**Comment by PVince81** [comment]
@jvillafanez I remove the integer lengths and the MySQL schema diffs are still the same.

Please rereview.

**Comment by jvillafanez** [comment]
I'd rather have the same datatype for all the timestamps. If so, 👍

**Comment by PVince81** [comment]
> I'd rather have the same datatype for all the timestamps

Same here... but that's how the table was: one is an integer and the other one is datetime, I can't easily change it here and it's not the purpose of this PR.

**Pull request merged**